### ISSUE TITLE
Custom autograd Functions are not aware of partial backward pass execution

### ISSUE THREAD

**Issue by mrapplexz** [pytorch/pytorch#174017]
### 🐛 Describe the bug

Hello! 

It seems that custom autograd Functions compute all the gradient edges pointing to tensors that `require_grad=True`, regardless computing these edges is required for a partial backward pass or not.

## Context

I am working on custom implementations of zero-bubble pipeline schedules. These schedules require that the gradient computation is done in two passes. During the first pass, we compute grad w.r.t. pipeline stage inputs. During the second pass, we compute grad w.r.t. model parameters. This allows us to minimize the time before gradients of pipeline inputs are sent to previous pipeline stage, therefore minimizing pipeline bubbles.

## What Works - Regular Ops

Regular pytorch ops work fine with partial backward pass. We can verify that using this code snippet:

```python
import torch
from torch import nn
from torch.profiler import profile, ProfilerActivity


class TestModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.param = nn.Parameter(torch.randn(128, 512))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x @ self.param

if __name__ == '__main__':
    with profile(
            activities=[
                ProfilerActivity.CPU,
                ProfilerActivity.CUDA
            ],
            record_shapes=True,
            with_stack=True
    ) as profiler:
        inputs = torch.randn((1, 128,), requires_grad=True)

        mod = TestModule()

        mod_outputs = mod(inputs)

        loss = mod_outputs.sum()

        # Do backward only with respect to the INPUTS, NOT WEIGHTS
        torch.autograd.backward(
            tensors=loss,
            inputs=inputs
        )
    profiler.export_chrome_trace("trace_aten.json")
```

After running the code, we can inspect the profile trace and see that the `aten::mm` is called exactly **once** during the backward pass.

<img width="751" height="226" alt="Image" src="https://github.com/user-attachments/assets/183af915-1b3a-446c-8888-5832a3758a35" />

## What Does not Work - Custom Functions

However, if we change matmul to a custom Function in the code snippet above, we get different results:

```python
import torch
from torch import nn
from torch.profiler import profile, ProfilerActivity


class TestMatmul(torch.autograd.Function):
    @staticmethod
    def forward(ctx, a, b):
        ctx.save_for_backward(a, b)
        return torch.matmul(a, b)

    @staticmethod
    def backward(ctx, grad_output):
        a, b = ctx.saved_tensors

        grad_a = grad_b = None

        if ctx.needs_input_grad[0]:
            grad_a = torch.matmul(grad_output, b.transpose(-1, -2))
        if ctx.needs_input_grad[1]:
            grad_b = torch.matmul(a.transpose(-1, -2), grad_output)

        return grad_a, grad_b


class TestModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.param = nn.Parameter(torch.randn(128, 512))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return TestMatmul.apply(x, self.param)


if __name__ == '__main__':
    with profile(
            activities=[
                ProfilerActivity.CPU,
                ProfilerActivity.CUDA
            ],
            record_shapes=True,
            with_stack=True
    ) as profiler:
        inputs = torch.randn((1, 128,), requires_grad=True)

        mod = TestModule()

        mod_outputs = mod(inputs)

        loss = mod_outputs.sum()

        # Do backward only with respect to the INPUTS, NOT WEIGHTS
        torch.autograd.backward(
            tensors=loss,
            inputs=inputs
        )
    profiler.export_chrome_trace("trace_custom.json")

```

If we inspect the profile trace, we can notice that `aten::mm` is called **twice**.

<img width="904" height="326" alt="Image" src="https://github.com/user-attachments/assets/378959b0-36cf-4725-aea6-4441f6b9c301" />

If we inspect the backward call using a debugger, we see that `ctx.needs_input_grad` is set to True for both input and parameter edges:

<img width="1546" height="253" alt="Image" src="https://github.com/user-attachments/assets/0cdf7130-bd34-4efe-b768-afd83b4abbfc" />

## The Hacks

We can "hack" this behavior by introducing additional checks in `backward(...)` implementation, like the DeepSeek team did for their DualPipe schedule:

https://github.com/deepseek-ai/DualPipe/blob/main/examples/example_dualpipe.py#L29

They introduce a singleton `WeightGradStore` object to control backward operations behavior externally, this fixes the problem but introduces unneeded complexity and screws up code dependency flow.  

## Expected Behavior

Computing only edges that **are required to be computed** during a partial backward pass is crucial for zero-bubble pipeline schedules. I expect that `torch.autograd.Function`s are aware of this scenario by `ctx.needs_input_grad` attributes of unneeded edges are set to False. 

I also guess that `torch.distributed.pipelining` API is affected by this, but I did not check it.

### Versions

```
Collecting environment information...
PyTorch version: 2.11.0.dev20251223+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.9 (main, Sep 10 2025, 23:06:05) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-51-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 570.124.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.13.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8462Y+
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   32
Socket(s):                            2
Stepping:                             8
CPU(s) scaling MHz:                   25%
CPU max MHz:                          4100.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5600.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hfi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
L1d cache:                            3 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             128 MiB (64 instances)
L3 cache:                             120 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110,112,114,116,118,120,122,124,126
NUMA node1 CPU(s):                    1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95,97,99,101,103,105,107,109,111,113,115,117,119,121,123,125,127
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy==1.19.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.0
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.28.9
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.11.0.dev20251223+cu128
[pip3] torchdata==0.11.0
[pip3] triton==3.6.0+git9844da95
[conda] Could not collect
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93

**Comment by mrapplexz** [comment]
I also created a workaround for this issue in my project  - https://github.com/d9d-project/d9d/commit/b3ce578ee2b36a490d9bbb73dd5ef11bc76dc1b5 - so if somebody will google a similar problem and find this issue, feel free to reuse that temporary solution :)  

Following the DeepSeek's DualPipe implementation, this one also introduces a global side-channel state manager (`GlobalGradContext`) that is configured in pipelining infra and is checked by all the custom Functions that are involved in both dI and dW passes (like Grouped GEMM in MoE layers). Global states are bad, but I guess it is the only option now.

**Comment by qflen** [comment]
Thanks for the clear repro and write-up @mrapplexz

@soulitzer alternative approach to the one above: refresh `THPFunction::needs_input_grad` once at the top of `PyNode::apply`, gated on `exec_info_` being non-empty, instead of a dynamic getter on ctx.needs_input_grad.

~ will PR tomorrow.
